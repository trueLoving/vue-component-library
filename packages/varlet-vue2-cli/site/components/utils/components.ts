import Vue, { ComponentOptions } from 'vue'
import { CombinedVueInstance } from 'vue/types/vue'

export function pickProps<T, U extends keyof T>(props: T, propsKey: U): T[U]
export function pickProps<T, U extends keyof T>(props: T, propsKey: U[]): Pick<T, U>
export function pickProps(props: any, propsKey: any): any {
  return Array.isArray(propsKey)
    ? propsKey.reduce((pickedProps: any, key) => {
        pickedProps[key] = props[key]
        return pickedProps
      }, {})
    : props[propsKey]
}

export interface MountComponentApi {
  instance: CombinedVueInstance<any, any, any, any, any>
  unmount(): void
}

export function mountComponent(component: ComponentOptions<Vue>, container = 'body', options = {}): MountComponentApi {
  const instance = new (Vue.extend(component))(options)
  const el = instance.$mount().$el
  const wrapper = document.querySelector(container)!
  wrapper.appendChild(el)

  return {
    instance,
    unmount() {
      instance.$destroy()
      wrapper.removeChild(el)
    },
  }
}
