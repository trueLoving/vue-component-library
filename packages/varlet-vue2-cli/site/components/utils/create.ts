import Vue, { ComponentOptions, RenderContext, VueConstructor } from 'vue'
import { SlotsMixin } from './mixins/slots'
import { ListenersMixin } from './mixins/listeners'

export type ComponentOptionsPlugin = ComponentOptions<Vue> & { install?: (app: VueConstructor) => void }

export function nextTick() {
  return new Promise(Vue.nextTick)
}

export function mergeSlots(context: RenderContext) {
  const scopedSlots = context.scopedSlots ?? context.data.scopedSlots ?? {}
  const slots = context.slots()

  Object.keys(slots).forEach((key) => {
    if (!scopedSlots[key]) {
      scopedSlots[key] = () => slots[key]
    }
  })

  return scopedSlots
}

export function defineComponent(sfc: ComponentOptionsPlugin & { name: string }) {
  if (sfc.mixins) {
    sfc.mixins.push(SlotsMixin)
    sfc.mixins.push(ListenersMixin)
  } else {
    sfc.mixins = [SlotsMixin, ListenersMixin]
  }

  return sfc
}
