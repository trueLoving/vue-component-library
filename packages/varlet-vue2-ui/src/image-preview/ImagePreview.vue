<template>
  <var-popup
    class="var-image-preview__popup"
    var-image-preview-cover
    transition="var-fade"
    :show="popupShow"
    :overlay="false"
    :close-on-click-overlay="false"
    :lock-scroll="lockScroll"
    :teleport="teleport"
    @open="$emit('open')"
    @close="$emit('close')"
    @closed="$emit('closed')"
    @opened="$emit('opened')"
    @route-change="$emit('route-change')"
  >
    <var-swipe
      class="var-image-preview__swipe"
      var-image-preview-cover
      :touchable="canSwipe"
      :indicator="indicator && images.length > 1"
      :initial-index="initialIndex"
      :loop="loop"
      @change="(index) => $emit('change', index)"
      v-bind="$attrs"
    >
      <template #default>
        <var-swipe-item
          class="var-image-preview__swipe-item"
          var-image-preview-cover
          v-for="image in images"
          :key="image"
        >
          <div
            class="var-image-preview__zoom-container"
            :style="{
              transform: `scale(${scale}) translate(${translateX}px, ${translateY}px)`,
              transitionTimingFunction,
              transitionDuration,
            }"
            @touchstart="handleTouchstart"
            @touchmove="handleTouchmove"
            @touchend="handleTouchend"
          >
            <img class="var-image-preview__image" :src="image" :alt="image" />
          </div>
        </var-swipe-item>
      </template>

      <template #indicator="{ index, length }">
        <slot name="indicator" :index="index" :length="length">
          <div class="var-image-preview__indicators" v-if="indicator && images.length > 1">
            {{ index + 1 }} / {{ length }}
          </div>
        </slot>
      </template>
    </var-swipe>

    <slot name="close-icon">
      <var-icon
        class="var-image-preview__close-icon"
        name="close-circle"
        var-image-preview-cover
        v-if="closeable"
        @click="close"
      />
    </slot>

    <div class="var-image-preview__extra">
      <slot name="extra" />
    </div>
  </var-popup>
</template>

<script>
import VarSwipe from '../swipe'
import VarSwipeItem from '../swipe-item'
import VarIcon from '../icon'
import VarPopup from '../popup'
import { defineComponent } from '../utils/create'
import { props } from './props'
import { toNumber } from '../utils/shared'

const DISTANCE_OFFSET = 12
const EVENT_DELAY = 200
const ANIMATION_DURATION = 200

export default defineComponent({
  name: 'VarImagePreview',

  components: {
    VarSwipe,
    VarSwipeItem,
    VarPopup,
    VarIcon,
  },

  inheritAttrs: false,

  props,

  data: () => ({
    popupShow: false,
    scale: 1,
    translateX: 0,
    translateY: 0,
    transitionTimingFunction: undefined,
    transitionDuration: undefined,
    canSwipe: true,
    startTouch: null,
    prevTouch: null,
    checker: null,
  }),

  computed: {
    initialIndex() {
      const { images, current } = this
      const index = images.findIndex((image) => image === current)
      return index >= 0 ? index : 0
    },
  },

  watch: {
    show: {
      handler(newValue) {
        this.popupShow = newValue
      },
      immediate: true,
    },
  },

  methods: {
    getDistance(touch, target) {
      const { clientX: touchX, clientY: touchY } = touch
      const { clientX: targetX, clientY: targetY } = target

      return Math.abs(Math.sqrt((targetX - touchX) ** 2 + (targetY - touchY) ** 2))
    },

    createVarTouch(touches, target) {
      return { clientX: touches.clientX, clientY: touches.clientY, timestamp: Date.now(), target }
    },

    zoomIn() {
      this.scale = toNumber(this.zoom)
      this.canSwipe = false
      this.prevTouch = null

      window.setTimeout(() => {
        this.transitionTimingFunction = 'linear'
        this.transitionDuration = '0s'
      }, ANIMATION_DURATION)
    },

    zoomOut() {
      this.scale = 1
      this.translateX = 0
      this.translateY = 0
      this.canSwipe = true
      this.prevTouch = null
      this.transitionTimingFunction = undefined
      this.transitionDuration = undefined
    },

    isDoubleTouch(currentTouch) {
      if (!this.prevTouch) {
        return false
      }

      return (
        this.getDistance(this.prevTouch, currentTouch) <= DISTANCE_OFFSET &&
        currentTouch.timestamp - this.prevTouch.timestamp <= EVENT_DELAY &&
        this.prevTouch.target === currentTouch.target
      )
    },

    isTapTouch(target) {
      if (!target || !this.startTouch || !this.prevTouch) {
        return false
      }

      return (
        this.getDistance(this.startTouch, this.prevTouch) <= DISTANCE_OFFSET &&
        (target === this.startTouch.target || target.parentNode === this.startTouch.target)
      )
    },

    handleTouchend(event) {
      this.checker = window.setTimeout(() => {
        this.isTapTouch(event.target) && this.close()
        this.startTouch = null
      }, EVENT_DELAY)
    },

    handleTouchstart(event) {
      this.checker && window.clearTimeout(this.checker)
      const { touches } = event
      const currentTouch = this.createVarTouch(touches[0], event.currentTarget)
      this.startTouch = currentTouch

      if (this.isDoubleTouch(currentTouch)) {
        this.scale > 1 ? this.zoomOut() : this.zoomIn()
        return
      }

      this.prevTouch = currentTouch
    },

    getZoom(target) {
      const { offsetWidth, offsetHeight } = target
      const { naturalWidth, naturalHeight } = target.querySelector('.var-image-preview__image')

      return {
        width: offsetWidth,
        height: offsetHeight,
        imageRadio: naturalHeight / naturalWidth,
        rootRadio: offsetHeight / offsetWidth,
        zoom: toNumber(this.zoom),
      }
    },

    getLimitX(target) {
      const { zoom, imageRadio, rootRadio, width, height } = this.getZoom(target)
      if (!imageRadio) {
        return 0
      }
      const displayWidth = imageRadio > rootRadio ? height / imageRadio : width
      return Math.max(0, (zoom * displayWidth - width) / 2) / zoom
    },

    getLimitY(target) {
      const { zoom, imageRadio, rootRadio, width, height } = this.getZoom(target)
      if (!imageRadio) {
        return 0
      }
      const displayHeight = imageRadio > rootRadio ? height : width * imageRadio
      return Math.max(0, (zoom * displayHeight - height) / 2) / zoom
    },

    getMoveTranslate(current, move, limit) {
      if (current + move >= limit) {
        return limit
      }

      if (current + move <= -limit) {
        return -limit
      }

      return current + move
    },

    handleTouchmove(event) {
      if (!this.prevTouch) {
        return
      }

      const target = event.currentTarget
      const { touches } = event
      const currentTouch = this.createVarTouch(touches[0], target)

      if (this.scale > 1) {
        const moveX = currentTouch.clientX - this.prevTouch.clientX
        const moveY = currentTouch.clientY - this.prevTouch.clientY

        const limitX = this.getLimitX(target)
        const limitY = this.getLimitY(target)

        this.translateX = this.getMoveTranslate(this.translateX, moveX, limitX)
        this.translateY = this.getMoveTranslate(this.translateY, moveY, limitY)
      }

      this.prevTouch = currentTouch
    },

    close() {
      if (this.scale > 1) {
        this.zoomOut()
        window.setTimeout(() => this.$emit('update:show', false), ANIMATION_DURATION)
        return
      }

      this.$emit('update:show', false)
    },
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../icon/icon';
@import '../swipe/swipe';
@import '../swipe-item/swipeItem';
@import '../popup/popup';
@import './imagePreview';
</style>
