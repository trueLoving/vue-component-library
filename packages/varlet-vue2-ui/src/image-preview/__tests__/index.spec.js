import ImagePreview from '../index'
import VarImagePreview from '../ImagePreview.vue'
import Vue from 'vue'
import { delay, trigger, triggerDrag } from '../../utils/jest'
import { mount } from '@vue/test-utils'

test('test image preview plugin', () => {
  Vue.use(ImagePreview)
  expect(Vue.component(VarImagePreview.name)).toBeTruthy()
})

test('test image preview functional show & close', async () => {
  const onOpen = jest.fn()
  const onOpened = jest.fn()
  const onClose = jest.fn()
  const onClosed = jest.fn()

  ImagePreview({
    onOpen,
    onOpened,
    onClose,
    onClosed,
  })

  await delay(200)
  expect(onOpen).toHaveBeenCalledTimes(1)
  await delay(300)
  expect(onOpened).toHaveBeenCalledTimes(1)
  expect(document.querySelector('.var-popup').style.display).toBe('')

  ImagePreview.close()

  await delay(20)
  expect(onClose).toHaveBeenCalledTimes(1)
  await delay(300)
  expect(onClosed).toHaveBeenCalledTimes(1)
  expect(document.querySelector('.var-popup')).toBeFalsy()
})

test('test image preview onChange callback', async () => {
  const onChange = jest.fn()

  ImagePreview({
    images: ['https://varlet-vue2.vercel.app/cat.jpg', 'https://varlet-vue2.vercel.app/cat2.jpg'],
    onChange,
  })
  await delay(200)

  const track = document.querySelector('.var-swipe__track')
  await triggerDrag(track, -100, 0)
  expect(onChange).toHaveBeenLastCalledWith(1)

  await triggerDrag(track, 100, 0)
  expect(onChange).toHaveBeenLastCalledWith(0)

  ImagePreview.close()
  await delay(16)
})

test('test image preview closeable', async () => {
  ImagePreview({
    closeable: true,
  })
  await delay(200)
  await trigger(document.querySelector('.var-image-preview__close-icon'), 'click')
  await delay(300)
  expect(document.querySelector('.var-popup')).toBeFalsy()
})

test('test image preview extra slot', async () => {
  const template = `
    <var-image-preview :images="images" :show.sync="show">
      <template #extra>
       <div>extra</div>
      </template>
    </var-image-preview>`

  const wrapper = mount({
    components: {
      [VarImagePreview.name]: VarImagePreview,
    },
    data: () => ({
      images: ['https://varlet-vue2.vercel.app/cat.jpg'],
      show: true,
    }),
    template,
  })

  expect(wrapper.html()).toMatchSnapshot()
  expect(wrapper.find('.var-image-preview__extra').isVisible()).toBe(true)

  wrapper.destroy()
})
