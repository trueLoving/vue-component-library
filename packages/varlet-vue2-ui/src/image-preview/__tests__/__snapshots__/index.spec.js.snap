// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test image preview extra slot 1`] = `
"<transition-stub name=\\"var-fade\\" class=\\"var-image-preview__popup\\">
  <div class=\\"var--box var-popup\\" style=\\"z-index: 2010;\\">
    <transition-stub name=\\"var-fade\\">
      <div var-image-preview-cover=\\"\\" class=\\"var-popup__content var-elevation--3 var-popup--center\\" style=\\"z-index: 2012;\\">
        <div class=\\"var-swipe var-image-preview__swipe\\" var-image-preview-cover=\\"\\">
          <div class=\\"var-swipe__track\\" style=\\"width: 0px; transform: translateX(0px); transition-duration: 300ms;\\">
            <div class=\\"var-swipe-item var-image-preview__swipe-item\\" style=\\"width: 0px; transform: translateX(0px);\\" var-image-preview-cover=\\"\\">
              <div class=\\"var-image-preview__zoom-container\\" style=\\"transform: scale(1) translate(0px, 0px);\\"><img src=\\"https://varlet-vue2.vercel.app/cat.jpg\\" alt=\\"https://varlet-vue2.vercel.app/cat.jpg\\" class=\\"var-image-preview__image\\"></div>
            </div>
          </div>
          <!---->
        </div>
        <!---->
        <div class=\\"var-image-preview__extra\\">
          <div>extra</div>
        </div>
      </div>
    </transition-stub>
  </div>
</transition-stub>"
`;
